velox/docs/develop/window.rst

```diff
@@ -205,3 +205,107 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
 
 Rows with NULL values do not participate in the frames of the other rows.
 
+
+Empty frames
+------------
+
+Its possible that certain rows have empty frames. Empty frames occur when :
+
+* Both frame_start and frame_end fall before the first partition row.
+
+  E.g. in frame *ROWS BETWEEN 5 PRECEDING and 2 PRECEDING* the first 2 rows
+  have both frame bounds before the first partition row.
+
+* Both frame_start and frame_end fall after the partition end row.
+
+  E.g. in frame *ROWS BETWEEN 2 FOLLOWING and 5 FOLLOWING* the last 2 rows
+  have both frame bounds beyond the last partition row.
+
+* frame_start > frame_end row (as frame range is defined from frame_start to
+  frame_end).
+
+  E.g. In frame *ROWS BETWEEN UNBOUNDED PRECEDING AND 2 PRECEDING* the intent
+  is to compute aggregation from the partition start row to 2 rows prior
+  the current one. However, for the first 2 rows the frameStart
+  (frame index 0 for unbounded preceding) is ahead of 2 preceding
+  (indices -2 and -1).
+
+* For frames like *ROWS BETWEEN 2 PRECEDING AND 5 PRECEDING* or
+  *ROWS BETWEEN 5 FOLLOWING AND 2 FOLLOWING*, frame_start > frame_end for
+  all rows. So all frames are empty.
+
+**Partial frames**
+
+Like illustrated in the examples above, rows could have partial window frames.
+
+A partial frame occurs when:
+
+- frame_start < frame_end (so the empty frame condition is not violated)
+- One frame end is within partition bounds and the other end outside of it.
+  In these cases either frame_start is before the first partition row while
+  frame_end within the partition, or frame_start is in the partition but
+  frame_end beyond it.
+
+For such frames the violating boundary is clamped to the first or last row
+of the partition.
+
+Partial frames usually follow empty frames in a sliding window.
+
+E.g. In frame *ROWS BETWEEN 5 PRECEDING AND 2 PRECEDING*, the first 2 rows have
+frame_start and frame_end before the first partition row, so they are empty.
+But from 3rd - 5th row, the 5 preceding frame_start bound is outside the
+partition, but 2 preceding frame_end is within the partition. So for these
+3 rows frame_start is clamped to the first partition row.
+
+Similarly for frame *ROWS BETWEEN 2 FOLLOWING AND 5 FOLLOWING*, the 3-5th rows
+have frame_start within the partition, but frame_end beyond. So they are partial
+frames. The last 2 rows have both bounds outside the partition and are empty frames..
+
+Empty and partial window frames can be visualized as below
+
+.. image:: images/empty_frames.png
+    :width: 600
+    :align: center
+
+Frames with constant frame bounds (like 2 preceding) have strict sliding behavior.
+So the empty frames, partial frames and valid frames cluster together and follow
+(or precede) each other.
+
+Adhoc frames that use column values for bounds can have
```

Comment:
Could we give an example of adhoc frames?

Reply:
Done.